src/ucs/memory/memtype_cache.c

```diff
@@ -56,16 +57,23 @@ static UCS_F_ALWAYS_INLINE ucs_memtype_cache_t *ucs_memtype_cache_get_global()
     ucs_memtype_cache_t *memtype_cache = NULL;
     ucs_status_t status;
 
-    if (!ucs_global_opts.enable_memtype_cache) {
+    if (ucs_global_opts.enable_memtype_cache == UCS_NO) {
         return NULL;
     }
 
     /* Double-check lock scheme */
-    if (ucs_unlikely(ucs_memtype_cache_global_instance == NULL)) {
+    if (ucs_unlikely(ucs_memtype_cache_global_instance == NULL) &&
+        !ucs_memtype_cache_failed) {
         /* Create the memtype cache outside the lock, to avoid a Coverity error
            of lock inversion with UCS_INIT_ONCE from ucm_set_event_handler() */
         status = UCS_CLASS_NEW(ucs_memtype_cache_t, &memtype_cache);
         if (status != UCS_OK) {
+            /* If we failed to create the memtype cache once, do not try again */
+            ucs_memtype_cache_failed = 1;
+            if (ucs_global_opts.enable_memtype_cache == UCS_YES) {
+                ucs_warn("failed to create memtype cache: %s",
+                         ucs_status_string(status));
```

Comment:
should it be hard error if user force to enable memtype cache?

Reply:
i don't think it should be a fatal error since we can still continue the program without memtype cache